### ISSUE TITLE
additional comment attached to a service request not notifying the right people

### ISSUE THREAD

**Issue by JChipault** [USGS-WiM/whispersservices_django#388]
So it looks like when someone makes comments in the service request section of an event that are NOT attached to the initial request, then the person that made that subsequent comment is the person getting the notification. Below is the notification that got sent to jchipault@usgs.gov when jchipault made a comment on the event. 

![image](https://user-images.githubusercontent.com/49287076/84780356-17b18600-afab-11ea-8c27-55adfa8f8a4e.png)

Second case study: when I was using the abilenetest account and made a subsequent service request comment on an event that abilenetest owned, the table that logs notification distribution indicated that only abilenetest received the notification. 

4149 | 6/9/2020 | abilenetest | abilenetest | FALSE | event | 170893 | TEST. TEST. NWHC Services Comment: Event 170893

**To Reproduce**
Clicking "Add Comment" on an already existing service request

![image](https://user-images.githubusercontent.com/49287076/84782040-2dc04600-afad-11ea-8c9f-8cd606a8959a.png)

**Expected behavior**

In the notifications spreadsheet, we indicate that this this notification "toggles between nwhc-epi@usgs or HFS AND user who made the request and event owner"

But that "toggling" feels complicated... especially because I'm not sure how the system would determine that user Barb Bodenstein is actually part of the epi team and thus a comment by her would count as an entry by the nwhc-epi@usgs.gov pool.

So maybe instead of toggling, we simplify this so that every time anyone makes a subsequent comment in the service request section that there is no toggling with regard to who receives which message in the back-and-forth banter, but instead all notifications go to (nwhc-epi OR HFS, depending on location)+event owner+person who made initial service request+anyone else who has commented on THAT service request.

If it feels possible to subtract from this pool of people only the person who did make the most recent comment, then that'd be great because then they aren't receiving a notification about a comment they just input. But if that makes it too complicated, then don't worry about that.

As an example,
If jchipault adds a service request comment on event 170896 that is owned by nbaertlein and nbaertlein put in the initial service request, then the notification for jchipault's comment would go to nwhc-epi@usgs.gov (assuming the location of the event is not in the HFS jurisdiction) + nbaertlein (as event owner and initiator of the request), and jchipault. Alternatively, it would go to nwhc-epi and nbaertlein if it feels do-able to program in that the person who just made the comment doesn't receive the notification.  

If nbaertlein comments back, then the notification would go to nwhc-epi@usgs.gov, nbaertlein, and jchipault. I'm guessing that in this instance, nbaertlein would stay in the mix because he's event owner and initiator of this service request, but again, if it's easy to remove from the pool of people who are notified the person who just made the last comment (even if they are event owner and/or initiated the request) that'd be great. They know they just commented so don't really need a notification about that.

**Comment by BlakeDraper** [comment]
@aaronstephenson this will be more in your court so transferring to you. Let me know if you need any of my support for this one.

**Comment by aaronstephenson** [comment]
So, I think this is a lot simpler that it may seem. I mistakenly assigned  the comment creator as the notification recipient in one case, when the recipient should be nwhc-epi or HFS (the 'else' block here: https://github.com/USGS-WiM/whispersservices_django/blob/notifications/whispersservices/serializers.py#L403).

The fix is simple. Implemented in 35ce7f97

**Comment by JChipault** [comment]
@BlakeDraper can you let me know when this fixed has been pushed to test?

**Comment by BlakeDraper** [comment]
@JChipault sure thing. It is a backend update so it won't be part of the regular emails I send you guys with new releases published. 

So that change addresses the bug that was causing the emails to go to the wrong recipient. But you also proposed a whole new approach in the comment above. If that is something you want to pursue, could you create a separate issue for it and we will label it for future development effort? We can't really include a change like that into the Effort 2 release we are trying to publish right now.

**Comment by JChipault** [comment]
@BlakeDraper has this change been pushed to test? Trying to determine when I should check it again.

**Comment by BlakeDraper** [comment]
@JChipault Yes, you can go ahead and test.

**Comment by JChipault** [comment]
On 7/1, I logged in as abilenetest and made 3 comments in the service request area for event 170873 (owned by JChipault)

On 7/2, I logged in as JChipault and made a comment on abilenetest's service request and on JChipault's service request for event 170873. 

On 7/2, I logged in as abilenetest and made a comment on JChipault's service request for event 170873.

In looking at the snapshot of the notifications table that Neil sent to me, it doesn't look like any of those comments triggered notifications to anyone.

Thoughts?

**Comment by aaronstephenson** [comment]
I have dug into this. The code works fine locally in my dev environment. The problem appears to be that celery on the server is not getting the requests. Not sure why, but I'll keep looking.

**Comment by JChipault** [comment]
Emails are coming through now (Hooray!) but they are all going to whispers@usgs.gov.... is that simply because it's in the test env and emails would be directed differently when we go live?

**Comment by aaronstephenson** [comment]
The source for all these Services Comment notifications that were created today were you and abilene test, so the recipient will be the epi user (see the code [here](https://github.com/USGS-WiM/whispersservices_django/blob/notifications/whispersservices/serializers.py#L406)). In the test environment, that user is the same as the WHISPers admin user, as set in the configuration table, which you can change at any time.

Or am I overlooking something?

**Comment by nbaertlein** [comment]
@aaronstephenson We do need to expand the notification list for non-EPI comments, and in looking at the code it looks like we could make a couple small tweaks to get us closer.

From what I can tell the code creates 4 user variables:
1.	Epi (HFS or Madison)
2.	Event Owner
3.	Service Requester
4.	Active commenter

If this is true could we simply add the event owner and service requester to the non-Epi notifications?  Ideally if the commenter was the event owner or service requester, we wouldn’t want them to be notified of something they just did.  Lastly, we think the source for the non-Epi notifications should be the commenter and not the service requester.

I know we’re out of time, so I attempted to modify your code to address these.  I’m obviously no python coder, but could either of the options in the attached work?  Please advise.

[whispers_service_request_comment_code.xlsx](https://github.com/USGS-WiM/whispersservices_django/files/5076387/whispers_service_request_comment_code.xlsx)

**Comment by aaronstephenson** [comment]
I can make these changes, it should only take a few minutes.

**Comment by aaronstephenson** [comment]
Implemented in 3ef1f1f0

**Comment by nbaertlein** [comment]
@aaronstephenson Thanks for fixing and implementing the new code.  I'm getting an error on some of the comments I'm making ("Error. Comment not saved. Error message: SyntaxError: Unexpected token T in JSON at position 0"), but the comment is still saved.  I ran a bunch of non-Epi comment scenarios and it seems to me the comments that fail are from users who were NOT the event owner or service requester.   Would it be possible the final else statement of the new code (lines 444-447) is causing the problem?  The 'list' function seems to be different than what's used in the other scenarios.

'# comment created by by neither service request creator nor service request event's creator
                else:
                    recipients = list({[epi_user.id, service_request.created_by.id,
                                        service_request.event.created_by.id, ]})
                    email_to = list({[epi_user.email, service_request.created_by.email,
                                      service_request.event.created_by.email, ]})

**Comment by aaronstephenson** [comment]
Sorry about that. Looks like I forgot to remove some code where I experimented with set literals. 😬
Fixed in 6e292c3e

**Comment by aaronstephenson** [comment]
It looks to me that Event 170907 did create notifications with the Madison Epi as the recipient, both a Service Request notification (ID 7911) and a Service Request Comment notification (ID 7912). It does not look like Event 170907 triggered either of those notifications, however.

After looking through the code, I suspect what happened was you created the service request for event 170907 after that event was created, and created the service request for event 170908 at the same time as that event, because the code that creates the service request notification was only in the Service Request endpoint, not in the Event endpoint. I remedied that in 64048899 so that the service request notification will get created in both situations. I just pushed this fix to the test environment.
